Extending/Packages/Good-practice/index.md

```diff
@@ -0,0 +1,161 @@
+
+As an example if you wanted to read a file called `robots.txt` from the `wwwroot` folder of a site in a controller you can use the `WebRootFileProvider` to get to the root of the site and read the file. 
+
+```cs
+public class MyController: UmbracoAuthorizedJsonController
+{
+    public MyController(IWebHostEnvironment webHostEnvironment)
+    {
+        var webRootProvider = webHostEnvironment.WebRootFileProvider;
+
+        var myFileInfo = webRootProvider.GetFileInfo("robots.txt");
+        if (myFileInfo.Exists)
+        {
+            using (var stream = myFileInfo.CreateReadStream())
+            {
+                using(var streamReader = new StreamReader(stream))
+                {
+                    var text = streamReader.ReadToEnd();
+                }
+            }
+        }
+    }
+}
+```
+
+:::note
+This is the preferred method for file IO because not all files served up by a site may in fact be in the wwwroot folder when you expect them to be - this is especially true if the site is using Razor class library to insert static files.
+:::
+
+### Path manipulation
+Building folder path strings manually can cause problems when swapping between file systems. Windows uses the backslash character to '\\' separate folders and files, Linux uses the forward slash '/'.
+
+You should use the .Net `Path` commands wherever possible when building paths to ensure that the correct path is built. 
+
+``` cs
+// WRONG: don't build paths manually.
+var myPath = webrootPath + "\robots.txt";
+
+// CORRECT: build the path using the correct separator for the current file system.
+var myPath = Path.Combine(webrootPath, "robots.txt"); 
+```
+
+If you do need to build a path manually use `Path.DirectorySeparatorChar` to get the correct separator for the filesystem.
+
+## Settings
+All but the most simple package will probably require some settings to be stored for the users to control and change the behavior of how the package works on a site. Where you store these settings will depend a lot on the nature of the package. 
+
+### Property Editors
+Property editors should store their settings as part of their datatype in Umbraco. This is the standard way property editors behavior is controlled, it is familar to users and is supported by deployment tools. 
+
+### Don't save to appsettings.json
+You should not alter `appsettings.json` in code. 
+
+Settings in aspnet core are merged from a number of different locations at runtime. You cannot guarantee that `appsettings.json` is the location that a setting is read from and your users may well not want certain settings in that file. You can read settings from the configuration but you cannot assume they have come from `appsettings.json`.
+
+### Settings locations
+There are many options for where you might save your settings and a lot will depend on the nature of your package: 
+
+#### Save to database 
```

Comment:
Maybe a section about the IKeyValueService in Umbraco to store simple settings. I use this in my RedirectsViewer package : https://github.com/dawoe/umbraco-redirectsviewer/blob/develop/Source/Our.Umbraco.RedirectsViewer/Controllers/UserGroupsApiController.cs#L89

Reply:
yeah - originally i did write a load of potential implementation options into the document, it got to big 😞 - so i went for streamlined, this document is more 'pointers' than actual implementation, not sure the balance is right, but maybe there should be other pages for this type of thing ?

Reply:
I agree that we should this document from becoming too big, but maybe we can add at least a single mention of this, e.g. 'Simple values can also be stored in the database using the `IKeyValueService`' and link that to the API docs?

Reply:
I've added a line about IKeyValueService , but i am not sure there is actually a documentation page to link to ?